tools/kagami/src/swarm.rs

```diff
@@ -31,166 +31,461 @@ const FILE_VALIDATOR: &str = "validator.wasm";
+    #[group(required = true, multiple = false)]
+    // FIXME: I haven't found a way how to share `image` and `build` options between `file` and
+    //        `dir` modes with correct grouping logic. `command(flatten)` doesn't work for it,
+    //        so it's hard to share a single struct with "base source options"
+    pub struct ModeFileSource {
+        /// Same as `--image` for `swarm dir` subcommand
+        #[arg(long)]
+        pub image: Option<String>,
+        /// Same as `--build` for `swarm build` subcommand
+        #[arg(long, value_name = "PATH")]
+        pub build: Option<PathBuf>,
+    }
+
+    #[cfg(test)]
+    mod tests {
+        use std::fmt::{Debug, Display, Formatter};
+
+        use clap::{ArgMatches, Command, Error as ClapError};
+
+        use super::*;
+
+        struct ClapErrorWrap(ClapError);
+
+        impl From<ClapError> for ClapErrorWrap {
+            fn from(value: ClapError) -> Self {
+                Self(value)
+            }
+        }
+
+        impl Debug for ClapErrorWrap {
+            fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+                Display::fmt(&self.0, f)
+            }
+        }
+
+        fn match_args(args_str: impl AsRef<str>) -> Result<ArgMatches, ClapErrorWrap> {
+            let cmd = Command::new("test");
+            let cmd = SwarmArgs::augment_args(cmd);
+            let matches = cmd.try_get_matches_from(
+                std::iter::once("test").chain(args_str.as_ref().split(' ')),
+            )?;
+            Ok(matches)
+        }
+
+        #[test]
+        fn works_in_file_mode() {
+            let _ = match_args("-p 20 file --build . --config-dir ./config sample.yml").unwrap();
+        }
+
+        #[test]
+        fn works_in_dir_mode_with_github_source() {
+            let _ = match_args("-p 20 dir --build-from-github swarm").unwrap();
+        }
+
+        #[test]
+        fn doesnt_allow_config_dir_for_dir_mode() {
+            let _ = match_args("-p 1 dir --build-from-github  --config-dir ./ swarm").unwrap_err();
+        }
+
+        #[test]
+        fn doesnt_allow_multiple_sources_in_dir_mode() {
+            let _ = match_args("-p 1 dir --build-from-github --build . swarm").unwrap_err();
+        }
+
+        #[test]
+        fn doesnt_allow_multiple_sources_in_file_mode() {
+            let _ = match_args("-p 1 file --build . --image hp/iroha --config-dir ./ test.yml")
+                .unwrap_err();
+        }
+
+        #[test]
+        fn doesnt_allow_github_source_in_file_mode() {
+            let _ =
+                match_args("-p 1 file --build-from-github --config-dir ./ test.yml").unwrap_err();
+        }
+
+        #[test]
+        fn doesnt_allow_omitting_source_in_dir_mode() {
+            let _ = match_args("-p 1 dir ./test").unwrap_err();
+        }
+
+        #[test]
+        fn doesnt_allow_omitting_source_in_file_mode() {
+            let _ = match_args("-p 1 file test.yml --config-dir ./").unwrap_err();
+        }
```

Comment:
These should be referenced in the `iroha2-docs` repo as example snippets.

Reply:
Created: https://github.com/hyperledger/iroha-2-docs/issues/353